### PR TITLE
FIP-55: PCV Guardian

### PR DESCRIPTION
**Pull request by Joeysantoro**
https://tribe.fei.money/t/fip-55-pcv-guardian/3744

TODO
- [x] integration test

### CODE REVIEW

File: proposals/dao/fip_55.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
+import { ethers } from 'hardhat';
+import chai, { expect } from 'chai';
+import CBN from 'chai-bn';
+import {
+  DeployUpgradeFunc,
+  NamedContracts,
+  SetupUpgradeFunc,
+  TeardownUpgradeFunc,
+  ValidateUpgradeFunc
+} from '../../types/types';
```
Comment (ditchfieldcaleb) [inline proposals/dao/fip_55.ts:10]: use shorthand

**Review by ditchfieldcaleb — CHANGES_REQUESTED** [review]
Needs an integration test or two, otherwise good.

**Comment by ditchfieldcaleb** [comment]
@Joeysantoro I think we should add in a check when withdrawing to a safe address *other than the DAOTimelock* that the source and destination underlying token is the same. Otherwise, we could accidentally burn tokens by transferring them into a pcv deposit that can't handle them.

**Comment by ditchfieldcaleb** [comment]
Also there's an e2e test on this that is failing that I'd like to see fixed before approving.

**Review by eswak — APPROVED** [review]

**Comment by eswak** [comment]
> @Joeysantoro I think we should add in a check when withdrawing to a safe address _other than the DAOTimelock_ that the source and destination underlying token is the same. Otherwise, we could accidentally burn tokens by transferring them into a pcv deposit that can't handle them.

We shouldn't be able to burn tokens this way, because deposits have a `withdrawERC20` function

**Comment by Joeysantoro** [comment]
> @Joeysantoro I think we should add in a check when withdrawing to a safe address _other than the DAOTimelock_ that the source and destination underlying token is the same. Otherwise, we could accidentally burn tokens by transferring them into a pcv deposit that can't handle them.

Interesting. I think we need to sync on this because at a glance its a good idea but it removes the flexibility to rely on withdrawERC20. I think we should put this check inside the `depositAfter` logic

File: proposals/dao/fip_55.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+export const setup: SetupUpgradeFunc = async (addresses, oldContracts, contracts, logging) => {
+  logging && console.log('No setup');
+};
+
+export const teardown: TeardownUpgradeFunc = async (addresses, oldContracts, contracts, logging) => {
+  logging && console.log('No teardown');
+};
+
+export const validate: ValidateUpgradeFunc = async (addresses, oldContracts, contracts) => {
+  const pcvGuardian: PCVGuardian = contracts.pcvGuardian as PCVGuardian;
```
Comment (ElliotFriedman) [inline proposals/dao/fip_55.ts:58]: We shouldn't need to cast to type of PCVGuardian as the artifact that is loaded from contracts should be correct.

File: test/integration/tests/pcv.ts [hunk 3]
```diff
@@ -53,6 +53,41 @@ describe('e2e-pcv', function () {
...
+      await pcvGuardian.withdrawToSafeAddress(
+        contractAddresses.rariPool8FeiPCVDeposit,
+        contractAddresses.feiDAOTimelock,
+        ethers.constants.WeiPerEther,
+        true,
+        false
+      );
+
+      expect(await contracts.rariPool8FeiPCVDeposit.paused()).to.be.true;
+      expect(await contracts.fei.balanceOf(contractAddresses.feiDAOTimelock)).to.be.bignumber.equal(
```
Comment (ElliotFriedman) [inline test/integration/tests/pcv.ts:84]: We should add test to see delta between starting and ending balance.

**Review by ditchfieldcaleb — APPROVED** [review]

**Pull request merged**